ostree-upload-server.py

```diff
@@ -213,19 +213,44 @@ def run(self):
 
         logging.info("Starting server on %d..." % self._port)
 
-        logging.debug("task completed callback %s", latest_task_complete);
+        logging.debug("task completed callback %s", latest_task_complete)
 
         def completed_callback(latest_task_complete):
-            logging.debug("task completed callback %s", latest_task_complete);
+            logging.debug("task completed callback %s", latest_task_complete)
             latest_task_complete[:] = [time()]
 
-        workers = Workers(self._repo, partial(completed_callback, latest_task_complete))
+        workers = Workers(partial(completed_callback, latest_task_complete))
         workers.start(task_list, self._workers)
 
-        def webapp_callback(task_name, filepath):
-            task_list.add_task(Task(task_name, filepath))
-
-        webapp = UploadWebApp(__name__, active_upload_counter, webapp_callback)
+        push_adapters = {}
+        config = SafeConfigParser(allow_no_value = True)
+        config.read('ostree-upload-server.conf')
+        for section in config.sections():
+            if not section.startswith('remote-'):
+                continue
+            remote_dict = dict(config.items(section))
+            remote_name = section.split('-')[1]
+            adapter_type = remote_dict['type']
+            if adapter_type in adapter_types:
+                logging.debug("setting up adapter {0}, type {1}".format(remote_name, adapter_type))
+                push_adapters[remote_name] = (adapter_types[adapter_type])(remote_name, remote_dict)
+            else:
+                logging.error("adapter {0}: unknown type {1}".format(remote_name, adapter_type))
+
+        def webapp_callback(task):
+            task_list.add_task(task)
+
+        users = None
+
+        if config.has_section('users'):
+            users = dict(config.items('users'))
+
+        webapp = UploadWebApp(__name__,
+                              users,
+                              self._repo,
+                              active_upload_counter,
+                              push_adapters,
+                              webapp_callback)
 
         http_server = WSGIServer(('', self._port), webapp)
```

Comment:
Yeah, that's fine. I might take another look at this setup again later, but I don't think there's anything wrong with this approach.